source/adapters/level_zero/v2/command_buffer.cpp

```diff
@@ -0,0 +1,185 @@
+
+  if (!DriverSupportsImmediateAppend) {
+    logger::error("Adapter v2 is used but "
+                  "the current driver does not support the "
+                  "zeCommandListImmediateAppendCommandListsExp entrypoint.");
+    std::abort();
+  }
+}
+
+} // namespace
+
+ur_exp_command_buffer_handle_t_::ur_exp_command_buffer_handle_t_(
+    ur_context_handle_t context, ur_device_handle_t device,
+    v2::raii::command_list_unique_handle &&commandList,
+    const ur_exp_command_buffer_desc_t *desc)
+    : commandListManager(
+          context, device,
+          std::forward<v2::raii::command_list_unique_handle>(commandList)),
+      isUpdatable(desc ? desc->isUpdatable : false) {}
+
+ur_result_t ur_exp_command_buffer_handle_t_::closeCommandList() {
+  // It is not allowed to append to command list from multiple threads.
+  std::scoped_lock<ur_shared_mutex> guard(this->Mutex);
+
+  // Close the command lists and have them ready for dispatch.
+  ZE2UR_CALL(zeCommandListClose, (this->commandListManager.getZeCommandList()));
+  return UR_RESULT_SUCCESS;
+}
+
+namespace ur::level_zero {
+
+ur_result_t
+urCommandBufferCreateExp(ur_context_handle_t context, ur_device_handle_t device,
+                         const ur_exp_command_buffer_desc_t *commandBufferDesc,
+                         ur_exp_command_buffer_handle_t *commandBuffer) try {
+  checkImmediateAppendSupport(context);
+
+  if (!context->getPlatform()->ZeMutableCmdListExt.Supported) {
+    throw UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  }
+
+  using queue_group_type = ur_device_handle_t_::queue_group_info_t::type;
+  uint32_t queueGroupOrdinal =
+      device->QueueGroup[queue_group_type::Compute].ZeOrdinal;
+  v2::raii::command_list_unique_handle zeCommandList =
+      context->commandListCache.getRegularCommandList(device->ZeDevice, true,
+                                                      queueGroupOrdinal, true);
+
+  *commandBuffer = new ur_exp_command_buffer_handle_t_(
+      context, device, std::move(zeCommandList), commandBufferDesc);
+  return UR_RESULT_SUCCESS;
+
+} catch (const std::bad_alloc &) {
+  return exceptionToResult(std::current_exception());
+}
+
+ur_result_t
+urCommandBufferRetainExp(ur_exp_command_buffer_handle_t hCommandBuffer) try {
+  hCommandBuffer->RefCount.increment();
+  return UR_RESULT_SUCCESS;
+} catch (const std::bad_alloc &) {
+  return exceptionToResult(std::current_exception());
+}
+
+ur_result_t
+urCommandBufferReleaseExp(ur_exp_command_buffer_handle_t hCommandBuffer) try {
+  if (!hCommandBuffer->RefCount.decrementAndTest())
+    return UR_RESULT_SUCCESS;
+
+  delete hCommandBuffer;
+  return UR_RESULT_SUCCESS;
+} catch (...) {
+  return exceptionToResult(std::current_exception());
+}
+
+ur_result_t
+urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t hCommandBuffer) try {
+  UR_ASSERT(hCommandBuffer, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(!hCommandBuffer->isFinalized, UR_RESULT_ERROR_INVALID_OPERATION);
```

Comment:
this is not thread safe, I think you should move this (and setting isFinalized to closeCommandList).